### PR TITLE
Change apiversion of CRD from v1beta1 to v1

### LINKED ISSUE #80: Move CRD from `v1beta1` to `v1`
When creating the CR, the API server warns about the deprecation of `CustomResourceDefinition` in `v1beta1` and unavailability from 1.22:
```
Warning: apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
```

I'd be happy to take this up unless there's a specific reason to keep it in `v1beta1`. 
/assign

### PR DESCRIPTION
**Pull request by MadhavJivrajani**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
The CRD example in sample-controller uses `v1beta1` as the apiversion and since that will be deprecated from 1.22, this PR updates the apiversion from `v1beta1` to `v1`. A summary of changes:

- Update hack/update-codegen.sh to generate code for group version v1
- Clean up v1alpha1 generated code files and correspondingly update the controller implementation to use v1
- Update CRDs (crd.yaml, crd-status-subresource.yaml) to match requirements of current release
	- Add `versions` field and structures `schema` as made mandatory in apiextensions/v1
	- Add annotation for api-approved.kubernetes.io since CRD is under *k8s.io domain
- Delete crd-validation.yaml since structured schema is mandatory in v1
- Update README
#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/sample-controller/issues/80

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml [hunk 1]
```diff
@@ -1,13 +1,35 @@
...
 kind: CustomResourceDefinition
 metadata:
   name: foos.samplecontroller.k8s.io
+  annotations:
+    "api-approved.kubernetes.io": "unapproved, experimental-only"
 spec:
   group: samplecontroller.k8s.io
-  version: v1alpha1
+  versions:
+    - name: v1
```
Comment (roycaihw) [inline staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml:10]: Why don't we keep the group version `v1alpha1`? In that way the generated files won't get changed

File: staging/src/k8s.io/sample-controller/README.md [hunk 2]
```diff
@@ -134,30 +134,17 @@ This feature is beta and enabled by default in v1.9.
...
-
 ## Subresources
 
-Custom Resources support `/status` and `/scale` subresources as a [beta feature](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#subresources) in v1.11 and is enabled by default.
-This feature is [alpha](https://v1-10.docs.kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#subresources) in v1.10 and to enable it you need to set the `CustomResourceSubresources` feature gate on the [kube-apiserver](https://kubernetes.io/docs/admin/kube-apiserver):
-
-```sh
---feature-gates=CustomResourceSubresources=true
-```
+Custom Resources support `/status` and `/scale` [subresources](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#subresources) from v1.16 in GA and are enabled by default.
```
Comment (roycaihw) [inline staging/src/k8s.io/sample-controller/README.md:142]: Both `CustomResourceValidation` and `CustomResourceSubresources` are GA. The feature gates have been removed in 1.18 so they cannot be disabled. Could you update the comment to remove the "and are enabled by default"?

Comment (MadhavJivrajani) [reply staging/src/k8s.io/sample-controller/README.md:142]: Oh sorry about that, will update the comment. Thanks!

File: staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml (hunk 1 above)
Comment (roycaihw) [reply staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml:10]: that's a good point, but I don't think these examples should be the place where we encourage CRD authors to develop stable API.

a CRD author should start with an alpha API, and gradually evolve the API into a stable one using [CRD versioning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/). This example misses that journey. 

another nit is this CRD has a `"api-approved.kubernetes.io": "unapproved, experimental-only"`, so probably leaving it as `v1alpha1` makes more sense.

File: staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml [hunk 3]
```diff
@@ -1,13 +1,35 @@
-apiVersion: apiextensions.k8s.io/v1beta1
+apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   name: foos.samplecontroller.k8s.io
+  annotations:
+    "api-approved.kubernetes.io": "unapproved, experimental-only"
```
Comment (roycaihw) [inline staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml:6]: perhaps append "; please get an approval from Kubernetes API reviewers if you're trying to develop a CRD in the *.k8s.io or *.kubernetes.io groups" to the annotation, and add a comment that links to https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2337-k8s.io-group-protection/README.md

so people who read this example and try to develop a CRD will know the next steps.

File: staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml (hunk 1 above)
Comment (MadhavJivrajani) [reply staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml:10]: > a CRD author should start with an alpha API, and gradually evolve the API into a stable one using CRD versioning. This example misses that journey.

Got it 👍🏻 
> another nit is this CRD has a "api-approved.kubernetes.io": "unapproved, experimental-only", so probably leaving it as v1alpha1 makes more sense.

iirc, the k8s.io annotation is a requirement of the API version itself and not the group version?

**Comment by roycaihw** [comment]
LGTM otherwise. Thanks!

**Comment by MadhavJivrajani** [comment]
Hey @roycaihw, thanks for the feedback! 
I added a small additional section to mention CRD versioning since I thought that would be useful to know, please let me know if the current changes look good. 😄

File: staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml [hunk 4]
```diff
@@ -1,10 +1,33 @@
...
+        # schema used for validation
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              properties:
+                deploymentName:
+                  type: string
+                image:
```
Comment (roycaihw) [inline staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml:25]: `image` isn't part of the [API](https://github.com/kubernetes/kubernetes/blob/22d120b4d68c8993f2782b9020b230e15f626c3c/staging/src/k8s.io/sample-controller/pkg/apis/samplecontroller/v1alpha1/types.go#L35-L39). Let's remove it

File: staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml [hunk 5]
```diff
@@ -1,10 +1,33 @@
...
+                  type: string
+                replicas:
+                  type: integer
+                  minimum: 1
+                  maximum: 10
```
Comment (roycaihw) [inline staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml:30]: I forgot if [pruning](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#field-pruning) works on `status`. Let me verify

Comment (roycaihw) [reply staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml:30]: yes, pruning works on `status`:

having a CR with status in etcd:
```
$ kubectl get foo my-foo -oyaml
apiVersion: samplecontroller.k8s.io/v1alpha1
kind: Foo
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"samplecontroller.k8s.io/v1alpha1","kind":"Foo","metadata":{"annotations":{},"name":"my-foo","namespace":"default"},"spec":{"deploymentName":"my-deployment","replicas":1},"status":{"availableReplicas":1}}
  creationTimestamp: "2021-05-12T18:07:50Z"
  generation: 1
  name: my-foo
  namespace: default
  resourceVersion: "537"
  uid: 920b53d2-baee-473d-9d25-bcf8dbdb8294
spec:
  deploymentName: my-deployment
  replicas: 1
status:
  availableReplicas: 1
```
edit the CRD to omit status in the schema
```
$ cat crd.yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
metadata:
  name: foos.samplecontroller.k8s.io
  # for more information on the below annotation, please see
  # https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/2337-k8s.io-group-protection/README.md
  annotations:
    "api-approved.kubernetes.io": "unapproved, experimental-only; please get an approval from Kubernetes API reviewers if you're trying to develop a CRD in the *.k8s.io or *.kubernetes.io groups"
spec:
  group: samplecontroller.k8s.io
  versions:
    - name: v1alpha1
      served: true
      storage: true
      schema:
        # schema used for validation
        openAPIV3Schema:
          type: object
          properties:
            spec:
              type: object
              properties:
                deploymentName:
                  type: string
                image:
                  type: string
                replicas:
                  type: integer
                  minimum: 1
                  maximum: 10
      # subresources for the custom resource
      subresources:
        # enables the status subresource
        status: {}
  names:
    kind: Foo
    plural: foos
  scope: Namespaced
```
read the CR again. Note the `status` is pruned
```
$ kubectl get foo my-foo -oyaml
apiVersion: samplecontroller.k8s.io/v1alpha1
kind: Foo
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"samplecontroller.k8s.io/v1alpha1","kind":"Foo","metadata":{"annotations":{},"name":"my-foo","namespace":"default"},"spec":{"deploymentName":"my-deployment","replicas":1},"status":{"availableReplicas":1}}
  creationTimestamp: "2021-05-12T18:07:50Z"
  generation: 1
  name: my-foo
  namespace: default
  resourceVersion: "537"
  uid: 920b53d2-baee-473d-9d25-bcf8dbdb8294
spec:
  deploymentName: my-deployment
  replicas: 1
```

Comment (roycaihw) [reply staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml:30]: please add the following to the schema in both `crd.yaml` and `crd-status-subresource.yaml`
```
            status:
              type: object
              properties:
                availableReplicas:
                  type: integer
```

**Comment by roycaihw** [comment]
we can LGTM this PR once the schema is fixed: https://github.com/kubernetes/kubernetes/pull/101871#discussion_r631275688, https://github.com/kubernetes/kubernetes/pull/101871#discussion_r631290933

ideally we should have integration tests that make sure the example CRD yaml works, but that might be out of the scope of this PR

**Comment by MadhavJivrajani** [comment]
Thanks! I didn't realise pruning works on `status`, learnt something new 😄

File: staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml [hunk 6]
```diff
@@ -1,13 +1,41 @@
...
+            spec:
+              type: object
+              properties:
+                deploymentName:
+                  type: string
+                replicas:
+                  type: integer
+                  minimum: 1
+                  maximum: 10
+          status:
```
Comment (roycaihw) [inline staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml:29]: the indentation misses two spaces. `status` should be at the same level as `spec`

Comment (MadhavJivrajani) [reply staging/src/k8s.io/sample-controller/artifacts/examples/crd-status-subresource.yaml]: oh i didn't realise it was off by 2 spaces, sorry about that, I've updated it.

**Comment by roycaihw** [comment]
/lgtm

/assign @munnerz
for approval

**Comment by MadhavJivrajani** [comment]
Thanks for the help and feedback! ❤️

**Comment by munnerz** [comment]
Thanks for the contribution @MadhavJivrajani and for reviewing @roycaihw!

/approve

**Pull request merged**